nopassword/forms.py

```diff
@@ -1,61 +1,133 @@
+        else:
+            site_name = domain = domain_override
+
+        url = '{}://{}{}?code={}&next={}'.format(
+            'https' if request.is_secure() else 'http',
+            domain,
+            resolve_url(login_url or settings.LOGIN_URL),
+            login_code.code,
+            login_code.next,
+        )
+
+        context = {
+            'domain': domain,
+            'site_name': site_name,
+            'code': login_code.code,
+            'url': url,
+        }
+
+        if extra_context:
+            context.update(extra_context)
+
+        self.send_login_code(login_code, context)
+
+    def send_login_code(self, login_code, context, **kwargs):
+        for backend in get_backends():
+            if hasattr(backend, 'send_login_code'):
+                backend.send_login_code(login_code, context, **kwargs)
+                break
+        else:
+            raise ImproperlyConfigured(
+                'Please add a nopassword authentication backend to settings, '
+                'e.g. `nopassword.backends.EmailBackend`'
+            )
+
+
+class LoginForm(forms.Form):
+    code = forms.ModelChoiceField(
+        label=_('Login code'),
+        queryset=models.LoginCode.objects.select_related('user'),
+        to_field_name='code',
+        widget=forms.TextInput,
+        error_messages={
+            'invalid_choice': _('Login code is invalid. It might have expired.'),
+        },
+    )
+
+    error_messages = {
+        'invalid_code': _("Unable to log in with provided login code."),
+    }
+
     def __init__(self, request=None, *args, **kwargs):
-        """
-        If request is passed in, the form will validate that cookies are
-        enabled. Note that the request (a HttpRequest object) must have set a
-        cookie with the key TEST_COOKIE_NAME and value TEST_COOKIE_VALUE before
-        running this validation.
-        """
+        super(LoginForm, self).__init__(*args, **kwargs)
+
         self.request = request
-        self.user_cache = None
-        super(AuthenticationForm, self).__init__(*args, **kwargs)
-        self.fields['username'].label = _(get_username_field().capitalize())
-
-    def clean(self):
-        username = self.cleaned_data.get('username')
-
-        if username:
-            self.user_cache = authenticate(**{get_username_field(): username})
-            if self.user_cache is None:
-                raise forms.ValidationError(
-                    self.error_messages['invalid_login'])
-            elif not isinstance(self.user_cache, LoginCode) and \
-                    not self.user_cache.is_active:
-                raise forms.ValidationError(self.error_messages['inactive'])
-        self.check_for_test_cookie()
-        return self.cleaned_data
-
-    def check_for_test_cookie(self):
-        if self.request and not self.request.session.test_cookie_worked():
-            raise forms.ValidationError(self.error_messages['no_cookies'])
-
-    def get_user_id(self):
```

Comment:
Backward incompatible: Removed method, as `user_cache` contains a login code, not a user.